File: facebook/src/test/java/com/facebook/MockSharedPreference.java
/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 *
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook;

import android.content.SharedPreferences;
import android.support.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class MockSharedPreference implements SharedPreferences {

    private final HashMap<String, Object> preferenceMap;

    private final MockEditor editor;

    public MockSharedPreference () {
        preferenceMap = new HashMap<>();
        editor = new MockEditor(preferenceMap);
    }

    @Override
    public Map<String, ?> getAll() {
        return preferenceMap;
    }

    @Nullable
    @Override
    public String getString(String key, @Nullable String defValue) {
        return (String) preferenceMap.getOrDefault(key, defValue);
    }

    @Nullable
    @Override
    public Set<String> getStringSet(String key, @Nullable Set<String> defValues) {
        return (Set<String>) preferenceMap.getOrDefault(key, defValues);
    }

    @Override
    public int getInt(String key, int defValue) {
        return (int) preferenceMap.getOrDefault(key, defValue);
    }

    @Override
    public long getLong(String key, long defValue) {
        return (long) preferenceMap.getOrDefault(key, defValue);
    }

    @Override
    public float getFloat(String key, float defValue) {
        return (float) preferenceMap.getOrDefault(key, defValue);
    }

    @Override
    public boolean getBoolean(String key, boolean defValue) {
        return (boolean) preferenceMap.getOrDefault(key, defValue);
    }

    @Override
    public boolean contains(String key) {
        return preferenceMap.containsKey(key);
    }

    @Override
    public Editor edit() {
        return editor;
    }

    @Override
    public void registerOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

    }

    @Override
    public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

    }

    public static class MockEditor implements Editor {

        private final Map<String, Object> preferenceMap;

        MockEditor(final Map<String, Object> map) {
            preferenceMap = map;
        }

        @Override
        public Editor putString(String key, @Nullable String value) {
            preferenceMap.put(key, value);
            return this;
        }

        @Override
        public Editor putStringSet(String key, @Nullable Set<String> values) {
            preferenceMap.put(key, values);
            return this;
        }

        @Override
        public Editor putInt(String key, int value) {
            preferenceMap.put(key, value);
            return this;
        }

        @Override
        public Editor putLong(String key, long value) {
            preferenceMap.put(key, value);
            return this;
        }

        @Override
        public Editor putFloat(String key, float value) {
            preferenceMap.put(key, value);
            return this;
        }

        @Override
        public Editor putBoolean(String key, boolean value) {
            preferenceMap.put(key, value);
            return this;
        }

        @Override
        public Editor remove(String key) {
            preferenceMap.remove(key);
            return this;
        }

        @Override
        public Editor clear() {
            preferenceMap.clear();
            return this;
        }

        @Override
        public boolean commit() {
            return true;
        }

        @Override
        public void apply() {

        }
    }
}

File: facebook-testutil/src/main/java/com/facebook/MockSharedPreference.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook

import android.annotation.TargetApi
import android.content.SharedPreferences
import android.os.Build

@TargetApi(Build.VERSION_CODES.N)
class MockSharedPreference : SharedPreferences {
  private val preferenceMap: HashMap<String, Any?> = HashMap()
  private val editor: MockEditor = MockEditor(preferenceMap)
  override fun getAll(): Map<String, *> = preferenceMap

  override fun getString(key: String, defValue: String?): String? {
    return preferenceMap.getOrDefault(key, defValue) as String?
  }

  override fun getStringSet(key: String, defValues: Set<String>?): Set<String>? {
    return preferenceMap.getOrDefault(key, defValues) as Set<String>?
  }

  override fun getInt(key: String, defValue: Int): Int {
    return preferenceMap.getOrDefault(key, defValue) as Int
  }

  override fun getLong(key: String, defValue: Long): Long {
    return preferenceMap.getOrDefault(key, defValue) as Long
  }

  override fun getFloat(key: String, defValue: Float): Float {
    return preferenceMap.getOrDefault(key, defValue) as Float
  }

  override fun getBoolean(key: String, defValue: Boolean): Boolean {
    return preferenceMap.getOrDefault(key, defValue) as Boolean
  }

  override fun contains(key: String): Boolean {
    return preferenceMap.containsKey(key)
  }

  override fun edit(): SharedPreferences.Editor = editor

  override fun registerOnSharedPreferenceChangeListener(
      listener: SharedPreferences.OnSharedPreferenceChangeListener
  ) = Unit
  override fun unregisterOnSharedPreferenceChangeListener(
      listener: SharedPreferences.OnSharedPreferenceChangeListener
  ) = Unit

  class MockEditor internal constructor(private val preferenceMap: MutableMap<String, Any?>) :
      SharedPreferences.Editor {
    override fun putString(key: String, value: String?): SharedPreferences.Editor {
      preferenceMap[key] = value
      return this
    }

    override fun putStringSet(key: String, values: Set<String>?): SharedPreferences.Editor {
      preferenceMap[key] = values
      return this
    }

    override fun putInt(key: String, value: Int): SharedPreferences.Editor {
      preferenceMap[key] = value
      return this
    }

    override fun putLong(key: String, value: Long): SharedPreferences.Editor {
      preferenceMap[key] = value
      return this
    }

    override fun putFloat(key: String, value: Float): SharedPreferences.Editor {
      preferenceMap[key] = value
      return this
    }

    override fun putBoolean(key: String, value: Boolean): SharedPreferences.Editor {
      preferenceMap[key] = value
      return this
    }

    override fun remove(key: String): SharedPreferences.Editor {
      preferenceMap.remove(key)
      return this
    }

    override fun clear(): SharedPreferences.Editor {
      preferenceMap.clear()
      return this
    }

    override fun commit(): Boolean = true

    override fun apply() = Unit
  }
}
